src/RRule.php

```diff
@@ -0,0 +1,225 @@
+    /** @var RecurrRule */
+    protected $rrule;
+
+    /** @var ArrayTransformer */
+    protected $transformer;
+
+    public function __construct(string $rule, DateTimeInterface $start)
+    {
+        $this->rrule = new RecurrRule($rule);
+        $this->transformer = new ArrayTransformer();
+
+        $this->startAt($start);
+        $this->limitTransformer(3);
+    }
+
+    public function isDue(DateTimeInterface $dateTime): bool
+    {
+        if ($dateTime < $this->rrule->getStartDate() || $this->isExpired($dateTime)) {
+            return false;
+        }
+
+        $nextDue = $this->getNextRecurrences($dateTime);
+        if (! $nextDue->valid()) {
+            return false;
+        }
+
+        return $nextDue->current() == $dateTime;
+    }
+
+    public function getNextDue(DateTimeInterface $dateTime): DateTimeInterface
+    {
+        if ($this->isExpired($dateTime)) {
+            return $this->rrule->getEndDate();
+        }
+
+        $nextDue = $this->getNextRecurrences($dateTime, false);
+        if (! $nextDue->valid()) {
+            return $dateTime;
+        }
+
+        return $nextDue->current();
+    }
+
+    public function isExpired(DateTimeInterface $dateTime): bool
+    {
+        if ($this->rrule->repeatsIndefinitely()) {
+            return false;
+        }
+
+        $end = $this->rrule->getEndDate();
+        $end = $end ?: $this->rrule->getUntil();
+
+        return $end !== null && $end < $dateTime;
+    }
+
+    /**
+     * Set the start time of this frequency
+     *
+     * @param DateTimeInterface $start
+     *
+     * @return $this
+     */
+    public function startAt(DateTimeInterface $start): self
+    {
+        $startDate = clone $start;
+        // When the start time contains microseconds, the first recurrence will always be skipped, as
+        // the transformer operates only up to seconds level. See also the upstream issue #155
+        $startDate->setTime($start->format('H'), $start->format('i'), $start->format('s'));
+
+        $now = new DateTime();
+        $diff = (int) $now->diff($startDate)->format('%R%y');
+        // The maximum time difference between the current time and the start time must not exceed 5 years,
+        // otherwise the transformer would spend forever generating all the recurrences from the start time
+        // till now pointlessly.
+        if ($diff < static::MAX_YEAR_DIFF) {
+            $startDate->setDate($now->format('Y'), $start->format('m'), $start->format('d'));
+        }
+
+        $this->rrule->setStartDate($startDate);
+
+        return $this;
+    }
+
+    /**
+     * Set the end time of this frequency
+     *
+     * @param DateTimeInterface $end
+     *
+     * @return $this
+     */
+    public function endAt(DateTimeInterface $end): self
+    {
+        $this->rrule->setEndDate($end);
+
+        return $this;
+    }
+
+    /**
+     * Get the RRULE as a string
+     *
+     * @return string
+     */
+    public function getRuleString(): string
```

Comment:
What is this method for?

Reply:
Is used by the schedule forms when saving the expression to the database or INI files.